blockprod/src/detail/block_maker.rs

```diff
@@ -92,21 +93,30 @@ impl BlockMaker {
         Ok(returned_accumulator)
     }
 
-    pub fn make_block(
+    pub async fn generate_block(
         &self,
-        current_tip_id: Id<Block>,
-        accumulator: &dyn TransactionAccumulator,
+        transactions: Transactions,
     ) -> Result<Block, BlockProductionError> {
-        // TODO: this isn't efficient. We have to create the header first, then see if it obeys consensus rules, then construct the full block
-        let current_time = self.time_getter.get_time();
-        let block = Block::new(
-            accumulator.transactions().clone(),
-            current_tip_id.into(),
-            BlockTimestamp::from_duration_since_epoch(current_time),
-            common::chain::block::ConsensusData::None,
-            BlockReward::new(vec![]), // TODO: define consensus and rewards through NetworkUpgrades
-        )?;
-        Ok(block)
+        self.chainstate_handle
+            .call({
+                let current_tip_id = self.current_tip_id;
+                let current_tip_height = self.current_tip_height;
+                let time_getter = self.time_getter.clone();
+                move |this| {
+                    this.generate_block(
```

Comment:
Why are we generating the block in the chainstate thread? Isn't this the job of the block maker thread?

Reply:
After talking tonight with Sam, I'll be moving the generation code back into `blockprod` and will try to refactor out what's needed for consensus calculations via chainstate queries rather than trying to directly access the index *within* chainstate.